### PR TITLE
Refactored: upload queue/base collection interactions

### PR DESCRIPTION
**Pull request by jsilvestre**
Overall the code (i.e. architecture and interactions between the modules) is simpler, but it's still complex because there are many cases to handle: conflict, cancel upload, errors, multiple uploads, how an upload is reflected across various clients, or broken binary detection.

I tried to explicit interactions by using methods with coherent names. Let me describe the architecture and how the interactions work.

- baseCollection: this is the collection with all models (File and Folder) that are loaded when the user browse.
- view: what is display to the user is basically a sub collection of the base collection (a folder). This sub collection is a Backbone.Projection
- upload queue: the upload queue is an async.queue which handles uploads 5 by 5. It used to be a Backbone.Collection, but not anymore. That being said, it still needs to know which models are in "upload state" mainly to output a feedback at the end of the upload. For that, I used another Backbone.Projection.

There are multiple states for a model during the upload process. States can be separate in two parts: local state and shared state.
* local state is maintained by the client performing the upload
* shared state is maintained across ALL clients through web sockets.
The local states and shared states are explicitely and broadly documented within `models/file`.

Shared state is needed so that when you upload a file, other clients see it and don't think it's broken (if it doesn't have its binary **yet** for instance).

Local state is needed to handle everything else: conflict, success, errors, ...
The local state is a model's property rather than an actual Backbone.Model attribute, to prevent it from being sent to the server. I could override what is sent to the server, but I figured it's cleaner like this.

Those states must be mutated and accessed through the getters and setters declared in `models/file`.

Once you understand how flows the data when it's added to the upload queue (by either addBlobs or addFolderBlobs), it's just a matter of understanding local state and shared state to display what you want in the view.

This architecture also means that whenever there is a bug in a feature, that can mean two things:
* the local state has been mutated in a way it should not have been.
* the view doesn't handle properly the states.

It's still complex because there are many cases to test and check whenever you change the upload queue code (or related code such as conflict management in `views/folder`). I don't think there is a magical solution to handle the complexity, we must put our effort into "hiding" the complexity behind a coherent API. My work is a step toward that goal, and I hope your comments will help me improve the PR.

I think it would be worth it to write unit tests for the upload queue, it's doable as long as we mock the actual "upload" which is easy, especially if I can manage to run it with NodeJS rather than the browser. What do you think?

### CODE REVIEW

**Comment by frankrousseau** [comment]
* Can you put your explanation in the wiki?
* Remove the notion of mutation, it's hard to understand. Talking about changes should be enough. What do you think?

**Comment by frankrousseau** [comment]
About tests, we still have to figure out how we can run tests from the command line.

File: client/app/collections/files.coffee [hunk 1]
```diff
@@ -88,8 +88,8 @@ module.exports = class FileCollection extends Backbone.Collection
                     @getFolderContent folder, ->
                         callback null, folder, collection
 
-    existingPaths: ->
-        @map (model) -> model.getRepository()
+    existingPaths: -> return @map (model) -> model.getRepository()
```
Comment (frankrousseau) [inline client/app/collections/files.coffee]: We should keep function on two lines. It's easier to maintain.

File: client/app/collections/upload_queue.coffee [hunk 2]
```diff
@@ -93,60 +103,87 @@ module.exports = class UploadQueue extends Backbone.Collection
...
+    # Process each element from the queue
+    uploadWorker: (model, next) =>
+        # Skip if there is an error.
+        if model.error
+            setTimeout next, 10
+
+        # If there is a conflict, the queue waits for the user to
+        # make a decision.
+        else if model.isConflict()
+            model.processOverwrite = @_decideOn.bind @, model, next
```
Comment (frankrousseau) [inline client/app/collections/upload_queue.coffee]: This line is too hard to understand.

File: client/app/collections/upload_queue.coffee [hunk 3]
```diff
@@ -93,60 +103,87 @@ module.exports = class UploadQueue extends Backbone.Collection
...
+        if choice
+            model.markAsUploading()
+            @_processSave model, done
+        else
+            model.resetStatus()
+            @remove model
+            done()
+
+
+    # Perform the actual persistence
```
Comment (frankrousseau) [inline client/app/collections/upload_queue.coffee]: This comment should explain how the persistence is performed.

File: client/app/collections/upload_queue.coffee [hunk 4]
```diff
@@ -93,60 +103,87 @@ module.exports = class UploadQueue extends Backbone.Collection
...
+    _processSave: (model, done) ->
+
+        # double check that we don't try to upload something we know will fail
+        if not model.isConflict() and not model.isErrored()
+            model.save null,
+                success: (model) ->
+                    model.file = null
+                    model.loaded = model.total
```
Comment (frankrousseau) [inline client/app/collections/upload_queue.coffee]: This line require explaination.

**Comment by jsilvestre** [comment]
* I would rather add that into a readme file, I feel like the odd we don't change the wiki are bigger than the odds we don't change a file in the code.
* We can remove the word "mutations" from the text without changing the meaning, the loss of accuracy is minimal. Thus I don't understand why it's harder to understand (but I am introduced to the concept of immutability so I think that's the reason it's okay for me to use "mutate").
* I was talking about running tests just on the queue and the model to ensure it is changed how it should be. I think that could be done in Node js only, like server-side tests, since the code is pretty portable once you remove the XHR interactions.

File: client/app/collections/upload_queue.coffee [hunk 5]
```diff
@@ -210,13 +258,12 @@ module.exports = class UploadQueue extends Backbone.Collection
...
                 # we can safely add files to bottom of queue
-                blobs = _.filter blobs, (blob) ->
-                    blob.name not in ['.', '..']
+                blobs = _.filter blobs, (blob) -> blob.name not in ['.', '..']
                 @addBlobs blobs, parent
                 return
 
             prefix = parent.getRepository()
-            parts = dir.split('/').filter (x) -> x # ?remove empty last part
+            parts = dir.split('/').filter (x) -> x # remove empty last part
```
Comment (frankrousseau) [inline client/app/collections/upload_queue.coffee]: this is elegant but weird and close to unreadable. Can we make a helper function of it?

File: client/app/lib/socket.coffee [hunk 6]
```diff
@@ -83,6 +82,9 @@ module.exports = class SocketListener extends CozySocketListener
             when 'update'
                 @collections.forEach (collection) =>
                     model = collection.get id
+
+                    return null if model?.isUploading()
```
Comment (frankrousseau) [inline client/app/lib/socket.coffee]: That return is ugly.

**Comment by frankrousseau** [comment]
Bug 1: When I upload a file (3GB) and cancels it, it's removed then added back.

**Comment by frankrousseau** [comment]
Bug 2: Success notification color is inconsistent (sometimes it's green, sometimes it's blue).

File: client/app/collections/upload_queue.coffee (hunk 2 above)
Comment (jsilvestre) [inline client/app/collections/upload_queue.coffee]: I'm feeling uneasy with explaining native function: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind. I'll try to do something though.

File: client/app/collections/upload_queue.coffee (hunk 5 above)
Comment (jsilvestre) [inline client/app/collections/upload_queue.coffee]: I didn't even understand it before trying hard this morning ahaha. It is very specific to folder's upload. I move it to heplers.

**Comment by frankrousseau** [comment]
Did your last commit fixes the bugs I mention above?

**Comment by jsilvestre** [comment]
* bug 1: did you wait before canceling? If so, how low did you wait?

* bug 2: I didn't manage to reproduce

**Comment by frankrousseau** [comment]
I waited like 10 seconds.

**Comment by frankrousseau** [comment]
bug 2 pops very often, you should try again to reproduce it. I think it happens when you upload several files at once.

**Comment by jsilvestre** [comment]
I temporarily removed the cancel upload feature, just to make sure everything works fine without it. Can you please try again @frankrousseau? I didn't manage to reproduce bug 2 a single time in almost a day of tries.
I fixed bug 3 though, which was conflict resolution didn't work in some cases.

**Comment by frankrousseau** [comment]
Thank you Joseph for that big change. I hope it will make maintenance easier. 

By the way can you handle two things now?

* Put back the cancel option 
* Add a timelapse of 20ms between the rendering of each line for list with more than 50 elements after the 50th element is displayed

**Comment by jsilvestre** [comment]
Sure, it was the plan! What about the documentation though: wiki or readme file in the repo?

**Comment by frankrousseau** [comment]
I prefer the wiki if you don't mind. The README is for quick reading and very precise information like how to modify the application.

**Comment by jsilvestre** [comment]
Okay, let's go for the wiki.

**Pull request merged**